From a62ae62d94dd5d1e015fc6acd978b2c24d42d442 Mon Sep 17 00:00:00 2001
From: Tormod Volden <debian.tormod@gmail.com>
Date: Sat, 10 Jan 2015 12:40:36 +0100
Subject: [PATCH] XRoar: Emulate Tormod's nx32 memory cartridge for Dragon

* This patch is for the XRoar emulator at http://www.6809.org.uk/xroar/ *
* It is based off the 0.33.2 release. *

EXTBANKS is the number of 32KB banks and can be 1, 4 or 16.

Writing a number to 0xFFBF activates the corresponding bank,
available at 0x8000-0xFFEF. Writing to 0xFFBE deactivates the
cartridge, and makes the internal ROM available again.

The cartridge is only activated in SAM map type 0 (it keeps
track of it by "listening" on 0xFFDE/F).

Note that the cartridge "firmware" is subject to change.

Signed-off-by: Tormod Volden <debian.tormod@gmail.com>
---
 src/machine.c | 33 +++++++++++++++++++++++++++++++++
 1 file changed, 33 insertions(+)

diff --git a/src/machine.c b/src/machine.c
index 56db6e0..80da258 100644
--- a/src/machine.c
+++ b/src/machine.c
@@ -60,6 +60,13 @@
 #include "xconfig.h"
 #include "xroar.h"
 
+/* number of 32KB banks in external memory cartridge: 1, 4 or 16 */
+#define EXTBANKS 4
+static uint8_t extmem[0x8000 * EXTBANKS];
+static _Bool extmem_map;	/* external memory mapped in or not */
+static _Bool extmem_ty;		/* tracks SAM ty register */
+static uint8_t extmem_bank;	/* which bank */
+
 unsigned int machine_ram_size = 0x10000;  /* RAM in bytes, up to 64K */
 uint8_t machine_ram[0x10000];
 static uint8_t *machine_rom;
@@ -943,6 +950,8 @@ void machine_reset(_Bool hard) {
 	}
 	if (hard) {
 		initialise_ram();
+		extmem_map = 0;
+		extmem_ty = 0;
 	}
 	mc6821_reset(PIA0);
 	mc6821_reset(PIA1);
@@ -1057,6 +1066,13 @@ static uint8_t read_byte(unsigned A) {
 	// produce a different "null" result on his 16K CoCo
 	if (SAM0->RAS)
 		CPU0->D = 0xff;
+
+	/* external memory cartridge */
+	if (A < 0xff00 && (A > 0xbfff || (!extmem_ty && extmem_map && A > 0x7fff))) {
+		CPU0->D = extmem[0x8000 * extmem_bank + (A & 0x7fff)];
+		SAM0->S = 8; /* decoder disabled */
+	}
+
 	switch (SAM0->S) {
 	case 0:
 		if (SAM0->RAS) {
@@ -1120,6 +1136,23 @@ static uint8_t read_byte(unsigned A) {
 }
 
 static uint8_t write_byte(unsigned A) {
+
+	/* external memory cartridge */
+	if ((A & ~1) == 0xffde) {
+		extmem_ty = A & 1;
+		LOG_PRINT("Switching to SAM map type %d\n", extmem_ty);
+	} else if ((A & ~1) == 0xffbe) {
+		extmem_map = A & 1;
+		extmem_bank = CPU0->D & (EXTBANKS - 1);
+		if (extmem_map)
+			LOG_DEBUG(3, "Switching to bank %d\n", extmem_bank);
+		else
+			LOG_DEBUG(3, "Switching off bank %d\n", extmem_bank);
+	} else if (A < 0xff00 && (A > 0xbfff || (!extmem_ty && extmem_map && A > 0x7fff))) {
+		extmem[0x8000 * extmem_bank + (A & 0x7fff)] = CPU0->D;
+		SAM0->S = 8; /* decoder disabled */
+	}
+
 	if ((SAM0->S & 4) || unexpanded_dragon32) {
 		switch (SAM0->S) {
 		case 1:
-- 
1.9.1

